packages/expo-updates/android/src/main/java/expo/modules/updates/db/BuildData.kt

```diff
@@ -0,0 +1,91 @@
+package expo.modules.updates.db
+
+import android.net.Uri
+import expo.modules.jsonutils.getNullable
+import expo.modules.updates.UpdatesConfiguration
+import expo.modules.updates.db.enums.UpdateStatus
+import org.json.JSONObject
+
+/**
+ * The build data stored by the configuration is subject to change when
+ * a user updates the binary.
+ *
+ * This can lead to inconsistent update loading behavior, for
+ * example: https://github.com/expo/expo/issues/14372
+ *
+ * This singleton wipes the updates when any of the tracked build data
+ * changes. This leaves the user in the same situation as a fresh install.
+ *
+ * So far we only know that `releaseChannel` and
+ * `requestHeaders[expo-channel-name]` are dangerous to change, but have
+ * included a few more that both seem unlikely to change (so we clear
+ * the updates cache rarely) and likely to
+ * cause bugs when they do. The tracked fields are:
+ *
+ *   UPDATES_CONFIGURATION_RELEASE_CHANNEL_KEY
+ *   UPDATES_CONFIGURATION_UPDATE_URL_KEY
+ *
+ * and all of the values in json
+ *
+ *   UPDATES_CONFIGURATION_REQUEST_HEADERS_KEY
+ */
+object BuildData {
+    var buildKey = "STATIC_BUILD_DATA"
+
+    fun ensureBuildDataIsConsistent(
+        updatesConfiguration: UpdatesConfiguration,
+        database: UpdatesDatabase,
+        scopeKey: String
+    ){
+        val buildJSON = getBuildData(database, scopeKey)
+        if (buildJSON == null || !isBuildDataConsistent(updatesConfiguration, buildJSON)) {
+            clearAllReadyUpdates(database)
+            setBuildData(updatesConfiguration, database, scopeKey)
+        }
+    }
+
+    fun clearAllReadyUpdates(database: UpdatesDatabase){
+        val allUpdates = database.updateDao().loadAllUpdatesWithStatus(UpdateStatus.READY)
```

Comment:
I'm unclear about the different states of updates and in general about the life cycle of embedded updates. I do know that if I clear all of them then I get strange undesirable behavior, curious if you think deleting `READY` updates is sufficient of this should be expanded.

Reply:
I would lean towards deleting everything in order to have a clean slate -- all updates and assets. (Leaving assets on disk unaccounted for in the DB currently means that they might never be deleted.) What is the strange behavior you're referring to here?

Reply:
When testing it locally, it seemed to wipe out even embedded update shipped with the new build. I redid this multiple times, but if that doesn't seem possible i can retry.

Reply:
I don't think it's possible for the embedded update to be cleared out of where the apk installs it on disk.

Reply:
hm... ok lemme rerun

Reply:
Yea upon zoom discussion and multiple trials on @esamelson setup as well as on a non emulated phone on mine, this seems like a local peculiarity of my emulator setup. Doing a universal update delete as suggested.